File: app/src/main/java/com/microlabs/trallet/BookDetailActivity.java
package com.microlabs.trallet;

import android.content.Intent;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.Toolbar;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.widget.AdapterView;
import android.widget.Spinner;
import android.widget.TextView;

import com.github.mikephil.charting.charts.PieChart;
import com.github.mikephil.charting.data.PieData;
import com.github.mikephil.charting.data.PieDataSet;
import com.github.mikephil.charting.data.PieEntry;
import com.github.mikephil.charting.utils.ColorTemplate;
import com.microlabs.trallet.adapter.CurrencySpinnerAdapter;
import com.microlabs.trallet.model.Currency;
import com.microlabs.trallet.model.Expense;

import java.text.NumberFormat;
import java.util.ArrayList;
import java.util.List;

import butterknife.BindView;
import butterknife.BindViews;
import butterknife.ButterKnife;
import butterknife.OnClick;
import io.realm.Realm;
import io.realm.RealmResults;

public class BookDetailActivity extends AppCompatActivity {

    @BindView(R.id.chart)
    PieChart chart;
    @BindView(R.id.lblFoodTotal)
    TextView lblFoodTotal;
    @BindView(R.id.lblTransportTotal)
    TextView lblTransportTotal;
    @BindView(R.id.lblShopTotal)
    TextView lblShopTotal;
    @BindView(R.id.lblOthersTotal)
    TextView lblOthersTotal;
    @BindViews({R.id.btnSeeExpense, R.id.lblTotal, R.id.lblFoodTotal, R.id.lblTransport, R.id.lblTransportTotal,
            R.id.lblShopTotal, R.id.lblShop, R.id.lblOthers, R.id.lblOthersTotal, R.id.lblAll, R.id.lblAllTotal, R.id.spinnerCurrency})
    List<View> views;
    @BindView(R.id.spinnerCurrency)
    Spinner spinnerCurrency;
    @BindView(R.id.lblAllTotal)
    TextView lblAllTotal;
    private int bookId;
    private String bookTitle;
    private ButterKnife.Setter<? super View, Integer> changeVisibility = new ButterKnife.Setter<View, Integer>() {
        @Override
        public void set(@NonNull View view, Integer value, int index) {
            view.setVisibility(value);
        }
    };
    private RealmResults<Currency> currencies;
    private double foodTotal;
    private double transportTotal;
    private double shopTotal;
    private double othersTotal;
    private double allTotal;
    private Realm realm;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_book_detail);
        ButterKnife.bind(this);
        Toolbar toolbar = (Toolbar) findViewById(R.id.toolbar);
        setSupportActionBar(toolbar);
        getSupportActionBar().setDisplayHomeAsUpEnabled(true);

        ButterKnife.apply(views, changeVisibility, View.GONE);

        bookId = getIntent().getIntExtra("bookId", 0);
        bookTitle = getIntent().getStringExtra("lblTitle");
        realm = Realm.getDefaultInstance();
        currencies = realm.where(Currency.class).findAll();

        setup();

        CurrencySpinnerAdapter adapterCurr = new CurrencySpinnerAdapter(this, R.layout.item_category, currencies);
        spinnerCurrency.setAdapter(adapterCurr);
        spinnerCurrency.setOnItemSelectedListener(new AdapterView.OnItemSelectedListener() {
            @Override
            public void onItemSelected(AdapterView<?> parent, View view, int position, long id) {
                Currency selectedItem = (Currency) spinnerCurrency.getSelectedItem();
                setCurrency(selectedItem.getName(), selectedItem.getValue());
            }

            @Override
            public void onNothingSelected(AdapterView<?> parent) {

            }
        });
    }

    @Override
    protected void onStart() {
        super.onStart();
        setup();
    }

    private void setup() {
        RealmResults<Expense> realmResults = realm.where(Expense.class).equalTo(Expense.fBook, bookId).findAll();
        List<PieEntry> entries = new ArrayList<>();
        foodTotal = calculateTotalExpense(realmResults.where().equalTo(Expense.fCategory, 1).findAll());
//        lblFoodTotal.setText(foodTotal + "");
        PieEntry entry2 = new PieEntry((float) foodTotal);
        entry2.setLabel("Food");
        entries.add(entry2);
        transportTotal = calculateTotalExpense(realmResults.where().equalTo(Expense.fCategory, 2).findAll());
//        lblTransportTotal.setText(transportTotal + "");
        PieEntry entry1 = new PieEntry((float) transportTotal);
        entry1.setLabel("Transport");
        entries.add(entry1);
        shopTotal = calculateTotalExpense(realmResults.where().equalTo(Expense.fCategory, 3).findAll());
//        lblShopTotal.setText(shopTotal + "");
        PieEntry entry = new PieEntry((float) shopTotal);
        entry.setLabel("Shop");
        entries.add(entry);
        othersTotal = calculateTotalExpense(realmResults.where().equalTo(Expense.fCategory, 4).findAll());
//        lblOthersTotal.setText(othersTotal + "");
        PieEntry entry3 = new PieEntry((float) othersTotal);
        entry3.setLabel("Others");
        entries.add(entry3);
        allTotal = foodTotal + transportTotal + shopTotal + othersTotal;
//        lblAllTotal.setText(allTotal + "");

        PieDataSet dataSet = new PieDataSet(entries, "Categories");
        dataSet.setColors(ColorTemplate.COLORFUL_COLORS);

        chart.setData(new PieData(dataSet));
        chart.setUsePercentValues(true);
        chart.invalidate();
    }

    private void setCurrency(String name, double value) {
        lblFoodTotal.setText(name + " " + NumberFormat.getInstance().format(foodTotal / value));
        lblTransportTotal.setText(name + " " + NumberFormat.getInstance().format(transportTotal / value));
        lblShopTotal.setText(name + " " + NumberFormat.getInstance().format(shopTotal / value));
        lblOthersTotal.setText(name + " " + NumberFormat.getInstance().format(othersTotal / value));
        lblAllTotal.setText(name + " " + NumberFormat.getInstance().format(allTotal / value));
    }

    @Override
    public boolean onPrepareOptionsMenu(Menu menu) {
        if (chart.getVisibility() == View.VISIBLE) {
            menu.getItem(0).setEnabled(false);
        } else {
            menu.getItem(1).setEnabled(false);
        }
        return super.onPrepareOptionsMenu(menu);
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.menu_book_detail, menu);
        return super.onCreateOptionsMenu(menu);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        if (item.getItemId() == R.id.menu_view_chart) {
            chart.setVisibility(View.VISIBLE);
            ButterKnife.apply(views, changeVisibility, View.GONE);
            invalidateOptionsMenu();
        } else if (item.getItemId() == R.id.menu_view_list) {
            chart.setVisibility(View.GONE);
            ButterKnife.apply(views, changeVisibility, View.VISIBLE);
            invalidateOptionsMenu();
        } else if (item.getItemId() == android.R.id.home) {
            finish();
        }
        return super.onOptionsItemSelected(item);
    }

    @OnClick(R.id.btnSeeExpense)
    public void onViewClicked() {
        Intent i = new Intent(BookDetailActivity.this, ExpenseActivity.class);
        i.putExtra("bookId", bookId);
        i.putExtra("lblTitle", bookTitle);
        startActivity(i);
    }

    private double calculateTotalExpense(List<Expense> expenses) {
        double total = 0;
        for (Expense expense : expenses) {
            total += expense.getValue() * currencies.where().equalTo(Currency.fId, expense.getCurrencyId()).findFirst().getValue();
        }
        return total;
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        realm.close();
    }
}

File: app/src/main/java/com/microlabs/trallet/BookDetailActivity.kt
package com.microlabs.trallet

import android.os.Bundle
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.view.View.GONE
import android.view.View.VISIBLE
import android.widget.AdapterView
import androidx.appcompat.app.AppCompatActivity
import androidx.databinding.DataBindingUtil
import com.microlabs.trallet.database.Currency
import com.microlabs.trallet.database.Expense
import com.microlabs.trallet.databinding.ActivityBookDetailBinding
import org.jetbrains.anko.startActivity
import java.text.NumberFormat

class BookDetailActivity : AppCompatActivity() {

    private val views: List<View> by lazy {
        listOf(binding.content.btnSeeExpense,
                binding.content.lblTotal, binding.content.lblFoodTotal, binding.content.lblTransport,
                binding.content.lblTransportTotal, binding.content.lblShopTotal, binding.content.lblShop,
                binding.content.lblOthers, binding.content.lblOthersTotal, binding.content.lblAll,
                binding.content.lblAllTotal, binding.content.spinnerCurrency)
    }
    private var bookId: Int = 0
    private var bookTitle: String = ""
    //    private var RCurrencies: RealmResults<RCurrency>? = null
    private var foodTotal: Double = 0.0
    private var transportTotal: Double = 0.0
    private var shopTotal: Double = 0.0
    private var othersTotal: Double = 0.0
    private var allTotal: Double = 0.0

    private lateinit var binding: ActivityBookDetailBinding
//    private val realm: Realm by lazy { Realm.getDefaultInstance() }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = DataBindingUtil.setContentView(this, R.layout.activity_book_detail)
        setSupportActionBar(binding.toolbar)
        supportActionBar!!.setDisplayHomeAsUpEnabled(true)

        views.forEach { it.visibility = GONE }

        bookId = intent.getIntExtra("bookId", 0)
        bookTitle = intent.getStringExtra("lblTitle")
//        RCurrencies = realm.where(RCurrency::class.java).findAll()

        setup()

//        val adapterCurr = CurrencySpinnerAdapter(this, R.layout.item_category, item_categoryRCurrencies!!)
//        spinnerCurrency.adapter = adapterCurr
        binding.content.spinnerCurrency.onItemSelectedListener = object : AdapterView.OnItemSelectedListener {
            override fun onItemSelected(parent: AdapterView<*>, view: View, position: Int, id: Long) {
                val selectedItem = binding.content.spinnerCurrency.selectedItem as Currency
                setCurrency(selectedItem.name, selectedItem.value)
            }

            override fun onNothingSelected(parent: AdapterView<*>) {

            }
        }

        binding.content.btnSeeExpense.setOnClickListener {
            startActivity<ExpenseActivity>("bookId" to bookId, "lblTitle" to bookTitle)
        }
    }

    override fun onStart() {
        super.onStart()
        setup()
    }

    private fun setup() {
//        val realmResults = realm.where(RExpense::class.java).equalTo(RExpense.fBook, bookId).findAll()
//        val entries = ArrayList<PieEntry>()
//        foodTotal = calculateTotalExpense(realmResults.where().equalTo(RExpense.fCategory, 1.toInt()).findAll())
//        //        lblFoodTotal.setText(foodTotal + "");
//        val entry2 = PieEntry(foodTotal.toFloat())
//        entry2.label = "Food"
//        entries.add(entry2)
//        transportTotal = calculateTotalExpense(realmResults.where().equalTo(RExpense.fCategory, 2.toInt()).findAll())
//        //        lblTransportTotal.setText(transportTotal + "");
//        val entry1 = PieEntry(transportTotal.toFloat())
//        entry1.label = "Transport"
//        entries.add(entry1)
//        shopTotal = calculateTotalExpense(realmResults.where().equalTo(RExpense.fCategory, 3.toInt()).findAll())
//        //        lblShopTotal.setText(shopTotal + "");
//        val entry = PieEntry(shopTotal.toFloat())
//        entry.label = "Shop"
//        entries.add(entry)
//        othersTotal = calculateTotalExpense(realmResults.where().equalTo(RExpense.fCategory, 4.toInt()).findAll())
//        //        lblOthersTotal.setText(othersTotal + "");
//        val entry3 = PieEntry(othersTotal.toFloat())
//        entry3.label = "Others"
//        entries.add(entry3)
//        allTotal = foodTotal + transportTotal + shopTotal + othersTotal
//        //        lblAllTotal.setText(allTotal + "");
//
//        val dataSet = PieDataSet(entries, "Categories")
//        dataSet.setColors(*ColorTemplate.COLORFUL_COLORS)
//
//        chart!!.data = PieData(dataSet)
//        chart!!.setUsePercentValues(true)
//        chart!!.invalidate()
    }

    private fun setCurrency(name: String, value: Double) {
        binding.content.lblFoodTotal.text = name + " " + NumberFormat.getInstance().format(foodTotal / value)
        binding.content.lblTransportTotal.text = name + " " + NumberFormat.getInstance().format(transportTotal / value)
        binding.content.lblShopTotal.text = name + " " + NumberFormat.getInstance().format(shopTotal / value)
        binding.content.lblOthersTotal.text = name + " " + NumberFormat.getInstance().format(othersTotal / value)
        binding.content.lblAllTotal.text = name + " " + NumberFormat.getInstance().format(allTotal / value)
    }

    override fun onPrepareOptionsMenu(menu: Menu): Boolean {
        if (binding.content.chart.visibility == VISIBLE) {
            menu.getItem(0).isEnabled = false
        } else {
            menu.getItem(1).isEnabled = false
        }
        return super.onPrepareOptionsMenu(menu)
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.menu_book_detail, menu)
        return super.onCreateOptionsMenu(menu)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when {
            item.itemId == R.id.menu_view_chart -> {
                binding.content.chart.visibility = VISIBLE
                views.forEach { it.visibility = GONE }
                invalidateOptionsMenu()
            }
            item.itemId == R.id.menu_view_list -> {
                binding.content.chart.visibility = View.GONE
                views.forEach { it.visibility = VISIBLE }
                invalidateOptionsMenu()
            }
            item.itemId == android.R.id.home -> finish()
        }
        return super.onOptionsItemSelected(item)
    }

    private fun calculateTotalExpense(expens: List<Expense>): Double {
        var total = 0.0
        for (expense in expens) {
//            total += expense.value * RCurrencies!!.where().equalTo(RCurrency.fId, expense.currencyId).findFirst()!!.value
        }
        return total
    }
}
